### DISCUSSION TITLE
Issues with lifetimes on custom AssetLoader

### DISCUSSION THREAD

**Discussion by Bekreth** [bevyengine/bevy#2162] [Q&A] [answered]
I've recently been attempting to setup a custom asset using serde to parse a YAML file similar to what was done in [this example](https://github.com/bevyengine/bevy/blob/main/examples/asset/custom_asset.rs#L9), but I've been running into quite a few issues getting the lifetimes correct and I'm hoping it is just my relative inexperience with Rust.

The YAML files I'm attempting to load bring in arbitrary String data, so I need my asset to have a borrowed `&str` variable, but this has lead to a whole host of issues.
```rust 
#[derive(Default, Debug, TypeUuid, Deserialize)]
#[uuid = "594838dc-63d6-4718-a223-2f76dc347a0c"]
pub struct GameConfig<'a> {
    pub window_name: &'a str,
}
#[derive(Default)]
pub struct YamlLoader<T: DeserializeOwned> {
    pub yamls: T,
}

impl<T> AssetLoader for YamlLoader<T>
where
    T: 'static + Send + Sync + Asset + DeserializeOwned + Debug,
{
    fn load<'a>(
        &'a self,
        bytes: &'a [u8],
        load_context: &'a mut LoadContext,
    ) -> BoxedFuture<'a, anyhow::Result<()>> {
        Box::pin(async move {
            serde_yaml::from_slice::<T>(bytes)
                .map_err(Error::msg)
                .map(|yaml_file| {
                    load_context.set_default_asset(LoadedAsset::new(yaml_file));
                })
        })
    }

    fn extensions(&self) -> &[&str] {
        &["yml", "yaml"]
    }
}

impl Plugin for YamlLoaderPlugin {
    fn build(&self, app: &mut bevy::prelude::AppBuilder) {
        app
            .add_asset::<GameConfig<'static>()
            .init_asset_loader::<YamlLoader<GameConfig<'static>>>();
    }
}

```

I've been aiming for this to be a generic YAML loader so any arbitrary struct can be loaded, but while attempting to attach lifetimes to my `Asset` I get the follow errors

```
indicate the anonymous lifetime: `<'_>`
...
implementation of `yaml::config::_::_serde::Deserialize` is not general enough
note: `yaml::config::GameConfig<'static>` must implement `yaml::config::_::_serde::Deserialize<'0>`, for any lifetime `'0`...
note: ...but `yaml::config::GameConfig<'_>` actually implements `yaml::config::_::_serde::Deserialize<'1>`, for some specific lifetime `'1`
```

I had a few questions around this:
1. How can I handle this anonymous lifetime
2. Can I assign a specific lifetime to an asset that isn't `'static`?
3. If I can't plug together Serde and Bevy within the AssetServer, are there any other recommendations for how to load arbitrary String data from file using Bevy?

**Comment by mockersf** [comment, marked as answer]
quick workaround: replace your `&'a str` by a `String`

**Reply by Bekreth** [reply]
I've tried that and run into the issue that Strings can't be moved because they don't implement `Copy` which means that the parent struct can't implement it either so it fails to meet `Asset` definition. I've tried a few variations of cloning the data without success.

**Reply by TheNeikos** [reply]
As far as I am aware, `Asset` does not require `Copy`, where did you meet that error?

**Reply by Bekreth** [reply]
Whoops, I was a bit lazy with my terminology there; its not the `Asset` that is giving me issues, its my attempt to convert my `Handle` into an `Asset` that pops up with errors
```rust
pub fn load_game_config(
    mut commands: Commands, 
    server: Res<AssetServer>,
    assets: Res<Assets<GameConfig>>,
) {
    let config_handle: Handle<GameConfig> = server.load("config.yaml");
    match assets.get(config_handle) {
        Some(config) =>  commands.insert_resource(config), // <------------- This is where I'm having issues
        None => panic!("expected data"),
    }
}
```
If I leave the `GameConfig` option as a borrowed value, I face
```
explicit lifetime required in the type of `assets`
```
Where, if I try to deference it and pass by value I get
```
cannot move out of a shared reference

move occurs because value has type `yaml::config::GameConfig`, which does not implement the `Copy` trait
```

**Reply by TheNeikos** [reply]
Right, that is because `Assets<T>::get` gives you a `&T` and if you want to insert a resource it needs to be `'static` (meaning not a reference or containing any non-static references). So an option here would be to clone it.

**Reply by bjorn3** [reply]
It is probably better to store the `Handle<GameConfig>` instead. Especially because the asset is likely not loaded yet when you call `assets.get()` here.

**Reply by Bekreth** [reply]
Thank you so much for the help, I was able to get this working by wrapping the `Handle<GameConfig>` into another struct!  While I was mucking around on this though, I ran into another small bug (feature?) wherein if two different systems are are made that load and then render an Asset respectively (like in the `example/asset/custom_assets.rs` example), the `Assets` object isn't populated if both of these systems are registered as startup systems.  I tried setting the various startup system to different stages without success.   Is this the intended behavior for `Assets`? I'd made the assumption that a `HandleId, Asset` pair would be loaded into the `Assets` structure in between stages, but this doesn't seem to be the case.

If this is the intended behavior, is there a best practice for pulling in Game wide system configurations that need to be present prior to launching the game? In my example here, it seems like it'd be rather wasteful to have the `WindowDescriptor` tied into a system that gets hit every tick when I only need to provide initialization data.  Is there a way to conditionally remove systems during runtime?

**Reply by bjorn3** [reply]
It can take a while for an asset to be loaded. Large assets can take seconds or minutes to load. While the asset is being loaded `assets.get()` returns `None`.
